tracer/src/Datadog.Trace/Activity/OperationNameMapper.cs

```diff
@@ -0,0 +1,156 @@
+// <copyright file="OperationNameMapper.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+using System;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.Activity
+{
+    /// <summary>
+    /// Helper class to map <see cref="SpanKinds"/> and various tags on an <c>Activity</c> or OpenTelemetry Span to a <see cref="Span.OperationName"/>.
+    /// </summary>
+    internal static class OperationNameMapper
+    {
+        internal static void MapToOperationName(Span span)
+        {
+            if (!string.IsNullOrEmpty(span.OperationName))
+            {
+                return; // if OperationName has a value it means there was an "operation.name" tag
+            }
+
+            string operationName;
+
+            if (span.Tags is not OpenTelemetryTags tags)
+            {
+                // while span.Tags can be something other than OpenTelemetryTags the
+                // ActivityHandlerCommon.StartActivity requires its "Tags" type to be
+                // OpenTelemetryTags to ensure that for all Activity objects that we
+                // process will be able to be mapped correctly.
+                return;
+            }
+
+            if (tags.IsHttpServer())
+            {
+                operationName = "http.server.request";
+            }
+            else if (tags.IsHttpClient())
+            {
+                operationName = "http.client.request";
+            }
+            else if (tags.IsDatabase())
+            {
+                operationName = $"{tags.DbSystem}.query";
+            }
+            else if (tags.IsMessaging())
+            {
+                operationName = $"{tags.MessagingSystem}.{tags.MessagingOperation}";
```

Comment:
I think the semantics that we are going off of from OpenTelemetry (still says their status is `Experimental`) have it that `messaging.operation` and `messaging.system` are required tags: https://opentelemetry.io/docs/specs/semconv/messaging/messaging-spans/#messaging-attributes

But after I got the `AzureServiceBus` stuff ported over yesterday I think I saw `messaging.system` tags with no `messaging.operation` and they led to `client.request` operation names. So maybe it would be preferable, but I'm not 100% sure what to do.

Reply:
A question for @zacharycmontoya perhaps then

Reply:
I think it makes sense for the other operation names to be `client.request`. `messaging.operation` should be one of a fixed, well-known set, so if a span doesn't have a corresponding operation then it's not really interacting with a message. I left comments below about the existing AzureServiceBus `"Message"` span where we may intelligently decorate it. Otherwise, `client.request` (low-cardinality) with different resource names seems fine to me.